pkg/io/buffered.go

```diff
@@ -0,0 +1,283 @@
+package io
+
+import (
+	"io"
+	"sync"
+
+	"github.com/pkg/errors"
+	"go.uber.org/atomic"
+)
+
+// BufferedReaderAt implements io.ReaderAt but extends and buffers reads up to the given buffer size.
+// Subsequent reads are returned from the buffers. Additionally it supports concurrent readers
+// by maintaining multiple buffers at different offsets, and matching up reads with existing
+// buffers where possible. When needed the least-recently-used buffer is overwritten with new reads.
+type BufferedReaderAt struct {
```

Comment:
i keep having thoughts about how to "improve" this reader, but they're all dependent on access patterns in parquet. i think we're just going to have to slowly tune/improve this based on our findings